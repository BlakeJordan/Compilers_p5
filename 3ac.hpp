#ifndef LAKE_3AC_HPP
#define LAKE_3AC_HPP

#include "list"
#include "map"
#include "err.hpp"
#include "symbol_table.hpp"

namespace lake{

class Procedure;
class IRProgram;

class Label{
public:
	Label(std::string nameIn){
		this->name = nameIn;
	}
	std::string toString(){
		return "lbl_" + this->name;
	}
private:
	std::string name;
};

class Opd{
public:
	virtual std::string toString() = 0;
};

class SymOpd : public Opd{
public:
	std::string toString() override{ 
		return mySym->getName();
	}
	const SemSymbol * getSym(){ return mySym; }
private:
	SymOpd(SemSymbol * sym) : mySym(sym) {} 
	SemSymbol * mySym;
	friend class Procedure;
	friend class IRProgram;
};

class LitOpd : public Opd{
public:
	LitOpd(std::string valIn) : val(valIn){ }
	std::string toString() override{
		return val;
	}
private:
	std::string val;
};

class AuxOpd : public Opd{
public:
	AuxOpd(std::string valIn) : val(valIn) { }
	std::string toString() override{
		return val;
	}
private:
	std::string val;
};

enum BinOp {
	ADD, SUB, DIV, MULT, OR, AND, EQ, NEQ, LT, GT, LTE, GTE
};
enum UnaryOp{
	NEG, NOT
};
enum Syscall {
	WRITE, READ, EXIT
};

class Quad{
public:
	Quad();
	void addLabel(Label * label);
	virtual std::string repr() = 0;
	std::string commentStr();
	virtual std::string toString(bool verbose=false);
	void setComment(std::string commentIn);
private:
	std::string myComment;
	std::list<Label *> labels;
};

class BinOpQuad : public Quad{
public:
	BinOpQuad(Opd * dstIn, BinOp opIn, Opd * src1In, Opd * src2In);
	std::string repr() override;
private:
	Opd * dst;
	BinOp op;
	Opd * src1;
	Opd * src2;
};

class UnaryOpQuad : public Quad {
public:
	UnaryOpQuad(Opd * dstIn, UnaryOp opIn, Opd * srcIn);
	std::string repr() override ;
private:
	Opd * dst;
	UnaryOp op;
	Opd * src;
};

class AssignQuad : public Quad{
	
public:
	AssignQuad(Opd * dstIn, Opd * srcIn)
	: dst(dstIn), src(srcIn)
	{ }
	std::string repr() override;

private:
	Opd * dst;
	Opd * src;
};

class LocQuad : public Quad {
public:
	LocQuad(Opd * srcIn, Opd * tgtIn);
	std::string repr() override;
private:
	Opd * src;
	Opd * tgt;
};

class JmpQuad : public Quad {
public:
	JmpQuad(Label * tgtIn);
	std::string repr() override;
private:
	Label * tgt;
};

class JmpIfQuad : public Quad {
public:
	JmpIfQuad(Opd * cndIn, bool invertIn, Label * tgtIn);
	std::string repr() override;
private:
	Opd * cnd;
	bool invert;
	Label * tgt;
};

class NopQuad : public Quad {
public:
	NopQuad();
	std::string repr() override;
};

class SyscallQuad : public Quad {
public:
	SyscallQuad(Syscall syscall, Opd * arg);
	std::string repr() override;
private:
	Opd * myArg;
	Syscall mySyscall;
};

class CallQuad : public Quad{
public:
	CallQuad(SemSymbol * calleeIn);
	std::string repr() override;
private:
	SemSymbol * callee;
};

class EnterQuad : public Quad{
public:
	EnterQuad(Procedure * proc);
	virtual std::string repr();
private:
	Procedure * myProc;
};

class LeaveQuad : public Quad{
public:
	LeaveQuad(Procedure * proc);
	virtual std::string repr() override;
private:
	Procedure * myProc;
};

class SetInQuad : public Quad{
public:
	SetInQuad(size_t indexIn, Opd * opdIn);
	std::string repr() override;
private:
	size_t index;
	Opd * opd;
};

class GetInQuad : public Quad{
public:
	GetInQuad(size_t indexIn, Opd * opdIn);
	std::string repr() override;
private:
	size_t index;
	Opd * opd;
};

class SetOutQuad : public Quad{
public:
	SetOutQuad(size_t indexIn, Opd * opdIn);
	std::string repr() override;
private:
	size_t index;
	Opd * opd;
};

class GetOutQuad : public Quad{
public:
	GetOutQuad(size_t indexIn, Opd * opdIn);
	std::string repr() override;
private:
	size_t index;
	Opd * opd;
};

class Procedure{
public:
	Procedure(IRProgram * prog, std::string name);
	void addQuad(Quad * quad);
	Quad * popQuad();
	IRProgram * getProg();
	lake::Label * makeLabel();

	void gatherLocal(SemSymbol * sym);
	void gatherFormal(SemSymbol * sym);
	SymOpd * getSymOpd(SemSymbol * sym);
	AuxOpd * makeTmp();

	std::string toString(bool verbose=false); 
	std::string getName();

	lake::Label * getLeaveLabel();
private:
	EnterQuad * enter;
	Quad * leave;
	Label * leaveLabel;

	IRProgram * myProg;
	std::map<SemSymbol *, SymOpd *> locals;
	std::list<Opd *> temps; 
	std::list<SymOpd *> formals; 
	std::list<Quad *> bodyQuads;
	std::string myName;
	size_t maxTmp;
};

class IRProgram{
public:
	Procedure * makeProc(std::string name);
	Label * makeLabel();
	Opd * makeString(std::string val);
	void gatherGlobal(SemSymbol * sym);
	SymOpd * getGlobal(SemSymbol * sym);

	std::string toString(bool verbose=false);
private:
	size_t max_label = 0;
	size_t str_idx = 0;
	std::list<Procedure *> procs; 
	HashMap<AuxOpd *, std::string> strings;
	std::map<SemSymbol *, SymOpd *> globals;
};

}

#endif 
