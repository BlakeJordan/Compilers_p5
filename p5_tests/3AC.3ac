[BEGIN GLOBALS]
[END GLOBALS]
[BEGIN f LOCALS]
b (formal)
a (local)
[END f LOCALS]
enter f
getin 1 b
a := b
iffalse a goto lbl_1
setout 1 4
goto lbl_0
goto lbl_2
lbl_1: nop
setout 1 12
goto lbl_0
lbl_2: nop
lbl_0: leave f
[BEGIN main LOCALS]
a (local)
b (local)
tmp0 (tmp)
tmp1 (tmp)
tmp2 (tmp)
[END main LOCALS]
enter main
b := a
a := a ADD 1
setin 1 True
call f
getout 1 tmp0
tmp1 := a ADD tmp0
a := tmp1
tmp2 := a LT 500
lbl_4: nop
lbl_5: nop
iffalse tmp2 goto lbl_5
goto lbl_4
lbl_3: EXIT

