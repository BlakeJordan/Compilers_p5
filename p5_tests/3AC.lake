int f(bool b) {
  bool a;
  a = b;
  if(a) {
    return 4;
  } else {
    return 12;
  }
}

int main() {
  int a;
  int @ b;
  @b = a;
  a++;
  while(a < 500) {
    a = a + f(true);
  }
}

