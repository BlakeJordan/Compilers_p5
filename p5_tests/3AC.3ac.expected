[BEGIN GLOBALS]
x
y
str_0 "wow good job"
[END GLOBALS]
[BEGIN f LOCALS]
b (formal)
n (formal)
a (local)
tmp0 (tmp)
tmp1 (tmp)
[END f LOCALS]
enter f
getin 1 b
getin 2 n
a := b
iffalse a goto lbl_1
tmp0 := 4 ADD n
setout 1 tmp0
goto lbl_0
goto lbl_2
lbl_1: nop
tmp1 := 12 SUB n
setout 1 tmp1
goto lbl_0
lbl_2: nop
lbl_0: leave f
[BEGIN g LOCALS]
tmp0 (tmp)
[END g LOCALS]
enter g
tmp0 := x LT y
iffalse tmp0 goto lbl_4
WRITE x
goto lbl_5
lbl_4: nop
READ y
lbl_5: nop
goto lbl_3
lbl_3: leave g
[BEGIN z LOCALS]
[END z LOCALS]
enter z
goto lbl_6
lbl_6: leave z
[BEGIN main LOCALS]
a (local)
b (local)
c (local)
tmp0 (tmp)
tmp1 (tmp)
tmp2 (tmp)
tmp3 (tmp)
tmp4 (tmp)
tmp5 (tmp)
tmp6 (tmp)
tmp7 (tmp)
tmp8 (tmp)
tmp9 (tmp)
tmp10 (tmp)
tmp11 (tmp)
tmp12 (tmp)
tmp13 (tmp)
tmp14 (tmp)
tmp15 (tmp)
tmp16 (tmp)
[END main LOCALS]
enter main
tmp0 @ b
tmp0 := a
a := a ADD 1
a := a SUB 1
tmp1 := NEG a
a := tmp1
lbl_8: nop
tmp2 := a LT 500
iffalse tmp2 goto lbl_9
tmp3 := 1 OR 0
setin 1 tmp3
setin 2 15
call f
getout 1 tmp4
tmp5 := a ADD tmp4
a := tmp5
goto lbl_8
lbl_9: nop
tmp6 := y DIV 6
tmp7 := x LTE tmp6
c := tmp7
tmp8 := y SUB 14
tmp9 := x GTE tmp8
c := tmp9
tmp10 := y MULT 15
tmp11 := x GT tmp10
c := tmp11
WRITE str_0
tmp12 @ b
READ tmp12
call g
call z
tmp13 := NOT c
tmp14 := 1 EQ 0
tmp15 := 1 AND tmp14
tmp16 := tmp13 NEQ tmp15
iffalse tmp16 goto lbl_10
setout 1 1
goto lbl_7
lbl_10: nop
setout 1 0
goto lbl_7
lbl_7: EXIT